Dev/incremental media model

agnessnowplow opened the pull request:
This is the Redshift only incremental model currently working together with the web package.

The  `_interactions_this_run` table follows the web model\`s incremental logic and takes `snowplow_web_base_events_this_run` as an input to then add the various contexts to enrich the base table with the additional media related fields.

The  `_plays_by_pageview_this_run` table aggregates the `_interactions_this_run`'table and serves as a basis for the incrementalised derived table `_plays_by_pageview`.

The main `_media_stats` derived table will also be updated incrementally based on the `_plays_by_pageview` derived table, on a play by pageview basis after a set time window. This is to prevent complex and expensive queries due to metrics which need to take the whole page_view events into calculation. This way the metrics will only be calculated once per pageview/media, after no new events are expected.

The additional `_pivot_base` table is there to calculate the percent_progress boundaries and weights that are used to calculate the total play time and other related media fields.

Once we are happy about this way of incrementalisation, I will update the schema names (only the media_stats and plays_by_pageview should be in the derived schema, the rest could go to scratch, to follow the conventions)

For testing, I ran dbt test which passed and I also compared the results with the drop and recompute model by taking 18-19-20 Jan data and running the incremental model with 1 day batches. I have adjusted some of the metrics calculations during this exercise so the two models diverged in parts at the moment.

emielver commented on packages.yml:
```diff
@@ -0,0 +1,5 @@
+  - package: dbt-labs/dbt_utils
+    version: [">=0.7.0", "<0.9.0"]
+  - package: snowplow/snowplow_web
+    version: [">=0.5.0", "<0.6.0"]
```
You can bump the web version to `0.6.0` if it's still compatible with everything

emielver commented on models/web/snowplow_media_plays_by_pageview_this_run.sql:
```diff
@@ -0,0 +1,112 @@
+{{
+  config(
+    materialized='table',
```
There's no need to define the materialisation as `table` here when you have that in the `dbt_project.yml` file already

emielver commented on models/web/snowplow_media_pivot_base.sql:
```diff
@@ -0,0 +1,42 @@
+{{
+  config(
+    materialized='table'
```
There's no need to define the materialisation as `table` here when you have that in the `dbt_project.yml` file already

agnessnowplow replied:
I`ll be changing it to view just to be in line with the web model from this perspective, too.

emielver commented on models/web/snowplow_media_media_stats.sql:
```diff
@@ -0,0 +1,266 @@
+{{
+config(
```
Indentation is off here

emielver commented on models/web/snowplow_media_media_stats.sql:
```diff
@@ -0,0 +1,266 @@
+{{
+config(
+materialized= 'incremental',
```
Is there a reason why we're using dbt incrementalisation here rather than the custom snowplow incrementalisation?

emielver commented on models/web/snowplow_media_interactions_this_run.sql:
```diff
@@ -0,0 +1,65 @@
+{{
+  config(
+    materialized='table',
```
There's no need to define the materialisation as `table` here when you have that in the `dbt_project.yml` file already

emielver commented on models/custom/snowplow_media_user_stats.sql:
```diff
@@ -0,0 +1,43 @@
+
+  from {{ ref("snowplow_media_session_stats") }}
+
+  group by domain_userid
```
For consistency it would be good to just use `group by 1` here

emielver commented on dbt_project.yml:
```diff
@@ -0,0 +1,36 @@
+    +materialized: table
+    web:
+      +tags: "snowplow_web_incremental"
+    custom:
```
Would you want to disable `custom` by default? Seems like the logical default value but up to you

agnessnowplow replied:
good point, did not consider this yet but it is the logical thing to do

emielver commented on dbt_project.yml:
```diff
@@ -0,0 +1,36 @@
+  snowplow__percent_progress_boundaries: [10, 25, 50, 75]
+  snowplow__valid_play_sec: 30
+  snowplow__complete_play_rate: 0.99
+  snowplow__max_media_pv_window: 2
```
Should you add a default entry for `snowplow__incremental_materialization`?

agnessnowplow replied:
I'd rather keep the variables at their intended package level for more clarity within scoping but good point, I have added a comment to the overview docs to leave the default materialisation intact when configuring the web package variables.

emielver reviewed (commented):
Just a few suggestions around simplifying and keeping things more consistent but it looks good! I also have a lingering question about the incremental functionality of the `media_stats` table but for the rest it's looking great!

bill-warner commented on models/web/snowplow_media_interactions_this_run.sql:
```diff
@@ -0,0 +1,65 @@
+    from {{ ref("snowplow_web_base_events_this_run") }} as e
+
+    inner join {{ source('atomic', 'com_snowplowanalytics_snowplow_media_player_event_1') }} as mpe
+    on mpe.root_id = e.event_id
```
You should be joining on the tstamps here too otherwise you risk introducing duplicates

bill-warner commented on models/web/snowplow_media_media_stats.sql:
```diff
@@ -0,0 +1,266 @@
+
+    from {{ ref("snowplow_media_plays_by_pageview") }} p
+
+    left join {{ this }} t
```
Is this effectively a cross join?

agnessnowplow replied:
As clarified, this will be moved to the where clause as eventually it is only used to act as a date filter. I realised I also need to change the filter field from last_play to max(start_tstamp) from the upstream model, as this way I will also include impressions, not just plays. Otherwise any new impressions that happened in any subsequent run without any plays will be double counted.

bill-warner commented on models/web/snowplow_media_plays_by_pageview_this_run.sql:
```diff
@@ -0,0 +1,112 @@
+
+)
+
+, retention_rate as (
```
Is it possible to calculate this with a case statement in the prep CTE?

agnessnowplow replied:
good point, despite the name it's a max calculation and does not have to come from the dedupe part so will be moved.

agnessnowplow commented:
Hi Emiel! I have left comments above for all the recommended changes and questions you or Will had and made modifications accordingly. I also added documentation within `snowplow_media_player_overview.md `and `README.md`. As I was going through the guide I also realised I need some slight changes, for instance I added a `selectors.yml` file. I always use the prod db for testing purposes because that has correct data with enough for sampling. I checked 18-20 Jan period when I have data in consecutive days for testing incremental runs with daily limits but once we are happy with the calculations I will do a full testing with all the data just in case. When testing I use it as intended, in a new dbt project importing it as a package and running it together with the web model. Just noticed I left scratch_agnes as schema name, will make another commit shortly to correct this.

emielver commented:
Hey Agnes, thanks for this and great job on everything so far!! I have added some small QOL suggestions with regards to the documentation, as well as directory conventions that we seem to be following in the other packages.  I also pointed out what I think is an issue with an average calculation in one of the custom models, and I'm still a bit confused by the `snowplow_media_player_media_stats` model's construction. Maybe we could have a chat about that before I finish the review but everything else should be actionable! And wrt schema stuff, you can also change the value in your dbt `profiles.yml` of the `schema` variable which will prefix `scratch` and `derived` with whatever you put, e.g. `dbt_agnes` so then you won't need to change the `dbt_project.yml` when testing locally!

emielver commented:
@agnessnowplow 
There's also some issues with building the pages. Can you add a empty `.nojekyll` file to the docs folder, and can you do a docs generate and copy over the `['catalog.json', 'index.html', 'manifest.json', 'run_results.json']` files from your target folder to your docs folder after the docs have been generated, in the same vein as https://github.com/snowplow/dbt-snowplow-mobile/tree/main/docs
Also you should be admin now so you can check/build the docs yourself as well!

emielver commented on models/web/sources.yml:
```diff
@@ -0,0 +1,14 @@
+sources:
+  - name: atomic
+    schema: atomic
+    database: snplow5
```
These should be:
```
    schema: "{{ var('snowplow__atomic_schema', 'atomic') }}"
    database: "{{ var('snowplow__database', target.database) }}"
```

emielver replied:
And can you add source table descriptions? Doesn't have to be as extensive as this but it's a good idea to have something like this
https://github.com/snowplow/dbt-snowplow-web/blob/main/models/base/src_base.yml

emielver commented on models/web/schema.yml:
```diff
@@ -0,0 +1,379 @@
+        description: '{{ doc("col_avg_percent_played") }}'
+      - name: complete_plays
+        description: '{{ doc("col_complete_plays") }}'
+  - name: snowplow_media_player_user_stats
```
You should remove this table definition and tests from here and add it to the `custom` folder

agnessnowplow commented on docs/markdown/snowplow_media_player_overview.md:
```diff
@@ -0,0 +1,287 @@
+
+> `snowplow__percent_progress_boundaries`: [10, 25, 50, 75]
+
+The default list of percent progress values. It needs to be aligned with the values being tracked by the tracker. It is worth noting that the more these percent progress boundaries are being tracked the more accurate the play time calculations become. Please note that tracking 100% is unnecessary as there is a separate `ended` event which the model equates to achieving 100% and it also gets inclu ...
```
sure, I added it as a source.

agnessnowplow commented on docs/markdown/snowplow_media_player_overview.md:
````diff
@@ -0,0 +1,287 @@
+To begin the synching process please run the following script:
+
+```bash
+dbt run -m snowplow_web.base snowplow_media_player --vars 'snowplow__start_date: date_when_media_player_tracking_started'
````
You mean this way? dbt run -m snowplow_web.base snowplow_media_player --vars 'snowplow__start_date: <<YOUR_FIELD_HERE>>' The variable needs to be this one in this case.

emielver replied:
Actually the convention is like this `<target.schema>`, so I guess it would be like this:
dbt run -m snowplow_web.base snowplow_media_player --vars 'snowplow__start_date: <date_when_media_player_tracking_starts>'

agnessnowplow replied:
I like this more, explanatory like how it was before plus making sure the user catches the value to overwrite due to the  < >

agnessnowplow commented on docs/markdown/snowplow_media_player_overview.md:
````diff
@@ -0,0 +1,287 @@
+    users:
+      enabled: false
+```
+Running it, however can still be achieved by running the selector as defined in the web model. In order for it to work, you can copy the selectors.yml file from the web package to your dbt project`s main directory.
````
I have made the changes, and also changed the wording, they can actually take it from the media_player package as I included the basic web model selector there, too.

agnessnowplow commented on docs/markdown/snowplow_media_player_overview.md:
```diff
@@ -0,0 +1,287 @@
+
+{% raw %}
+
+# Snowplow Media Player Package
```
I`ll add this to the bottom of the intro, though ideally I would choose a smaller character: Please note that the package is not compatible with the [Flutter tracker](https://docs.snowplowanalytics.com/docs/collecting-data/collecting-from-own-applications/flutter-tracker/) as it is reliant on the Snowplow [JavaScript tracker](https://docs.snowplowanalytics.com/docs/collecting-data/collecting-from-own-applications/javascript-trackers/javascript-tracker/javascript-tracker-v3) as mentioned above.

agnessnowplow replied:
reading it back it sounds a bit passive aggressive don't you think? :)

emielver replied:
No I think it should be okay, although yeah could maybe remove "as mentioned above"

agnessnowplow commented on docs/markdown/snowplow_media_player_col_definitions.md:
```diff
@@ -0,0 +1,235 @@
+{% enddocs %}
+
+{% docs col_play_rate %}
+Total plays divided by impressions.
```
Good point, I'll add this clarification to the definition: `Please note that as the base for media plays is pageview / media_id, in case the same video is played multiple times within the same pageview, it will still count as one play.`

agnessnowplow commented on models/web/schema.yml:
```diff
@@ -0,0 +1,379 @@
+version: 2
```
Yeah, it makes sense. I also added a `snowplow_media_custom.yml` to the custom folder.

agnessnowplow commented on docs/markdown/snowplow_media_player_col_definitions.md:
```diff
@@ -0,0 +1,235 @@
+{% enddocs %}
+
+{% docs col_duration %}
+Total length of media in seconds.
```
It is the former. I have added your illustration to the definition as an example to make it clear. `play_time_sec` is the time it took to play the media file.

agnessnowplow commented on models/web/snowplow_media_player_plays_by_pageview.sql:
```diff
@@ -0,0 +1,12 @@
+{{
+  config(
+    materialized='view',
```
We could save storage space for something they might not even be using as it is essentially the same as the snowplow_media_player_base with a small filter on that they can just query as and when needed or use downstream. It was originally Will's idea to separate a view from there just so people will not get confused behind the underlying data.

agnessnowplow commented on models/custom/snowplow_media_player_session_stats.sql:
```diff
@@ -0,0 +1,53 @@
+
+)
+
+select
```
True, I usually specify all the fields by default just so I have them ready for docs specifications and any further transformations with the model. Will remove it for simplicity now.

agnessnowplow commented on models/web/scratch/snowplow_media_player_interactions_this_run.sql:
```diff
@@ -0,0 +1,65 @@
+    {{ dbt_utils.surrogate_key(['e.page_view_id', 'coalesce(y.player_id, me.html_id)' ]) }} play_id,
+    round(mp.duration) as duration,
+    case when me.media_type = 'audio' then 'audio' else 'video' end as media_type,
+    coalesce(y.schema_vendor||'-'||y.schema_name, me.schema_vendor||'-'||me.schema_name)  as media_player_type,
```
I like the idea of this macro and we can definitely add something like this to utils but I don't think it is needed here, I'd rather have the full value as null indicating that something is off than a half-done one. This is not part of any key and also not being filtered on so it is safe to leave null. If I'm not mistaken both should be populated by default.

agnessnowplow commented on models/custom/snowplow_media_player_user_stats.sql:
```diff
@@ -0,0 +1,43 @@
+
+  select
+    domain_userid,
+    avg(case when (video_plays + audio_plays) > 0 then avg_play_time_min end) as avg_play_time_min,
```
Good point. Now it is a question of the usual accuracy vs table scan / db cost, because if I wanted to make this accurate I'd have to add a new cte and calculate the average using the snowplow_media_player_base table. Perhaps the solution is to rename this field to `avg_session_play_time_min` to keep the content accurate while keeping the simplicity, not sure it would bother the users that much. What do you think?

agnessnowplow commented:
I have run the model within dev1 so that I can replace the docs for the site (previously it was all within one schema). I have also added more fields for interactions_this_run that are not used in the model but users might want to take advantage of them in their custom models. I also made a correction within base_this_run because the retention_rate was not taking the correct value in case the percent_progress was calculated from the ended event.

emielver commented on models/web/scratch/snowplow_media_player_base_this_run.sql:
```diff
@@ -64,21 +76,35 @@ with prep as (
     on i.play_id = d.play_id
 
-    where i.event_type in ('percentprogress') and (i.start_tstamp <= d.first_seek_time or d.first_seek_time is null)
+    where i.percent_progress is not null and (i.start_tstamp <= d.first_seek_time or d.first_seek_time is null)
```
This is changed because `percent_progress` could still be null when `event_type` is `percentprogress`?

agnessnowplow replied:
It changed because the previous solution was incomplete, not including 100 which comes from the `ended` event. (It was changed from NULL to 100 within the percent_progress field by the model itself during the interactions_this_run to account for 100% automatically). So I picked the wrong field originally. The other solution could have been to include ('percentprogress', 'ended') but I found it cleaner this way cause then I'd have had to put a case when statement as well.

emielver reviewed (commented):
On two `dbt run` commands and a `dbt test` I get two failed `unique` tests for `event_id` in `snowplow_media_player_interactions_this_run` and `snowplow_web_base_events_this_run`. Is this something you encountered? If this is resolved I'm happy to give the go ahead!

emielver reviewed (approved):
Errors in previous comment where due to duplicate source data, model seems to work well!! Great job on this!

agnessnowplow commented:
Thanks for your review! I fixed the bugs I came across: as discussed separately, I changed the ephemeral models to tables, including the docs so that the manifest table gets updated and the incremental runs don't get stuck. I also changed the full outer join back to account for the scenario when there is no data for the incremental run, which should preserve any pre-existing percent progress values from previous runs.

emielver reviewed (changes requested):
Just add the sort and dist keys and then it should be okay!

emielver commented on dbt_project.yml:
```diff
@@ -0,0 +1,41 @@
+    custom:
+      +schema: "scratch"
+      +tags: "snowplow_web_incremental"
+      +enabled: false
```
This enabled should be triggered by a variable, which should be defined in your vars and in the documentation so that users can easily enable/disable these custom tables, right? Or how else would we expect users to build off of them?

emielver replied:
Fair enough, although I do think we should change it to being behind a variable once this dbt bug is fixed https://github.com/dbt-labs/dbt-core/issues/3698 to align with the web and mobile packages and their optional modules, but it's good for now!

emielver reviewed (approved):
LGTM

emielver reviewed (approved):
LGTM!
